### DISCUSSION TITLE
Возможность объединения сотрудников

### DISCUSSION THREAD

**Discussion by Igralino** [bakhirev/assayo#30] [Ideas]
Хотелось бы получить возможность смердживать несколько учетных записей сотрудников в одну. 

В данный момент, насколько я понял, сотрудником является уникальная почта. При этом достаточно часто бывает ситуация, что у одного человека указано несколько почт (одна в локальном git config, другая в Gitlab/Github, третья на другом девайсе). Из-за этого статистика бьется на несколько сущностей с одинаковыми/похожими ФИО. Решением могло бы быть ручное объединение учетных записей в одну через интерфейс сайта

**Comment by fdman-liis** [comment]
see readme, "how to concat authors"

**Comment by bakhirev** [comment]
**Такой функционал был год** назад и от него отказались. Уникальным считается ИМЯ (у нас) и ПОЧТА (для гита, стата почту игнорирует) Далее по пунктам опишу, почему это произошло:

1. Люди пишут разные почты и разные имена. Авто-склейка в лоб не сработала.
2. Эксперимент с авто-склейкой по частичному совпадению (для diff < n%) так же провалился, т.к. в тестовой выборке были два брата с одной фамилией и отчеством. Разница между ними была меньше 5% (имя задано одной буквой, как "Иванов И.С." и "Иванов К.С."). Так же этот путь не адаптируется под азию (Китайская локаль), т.к. там 1 символ = 1 смыслу и имена могут сильно совпадать.
3. Далее сделали простой интерфейс, чтобы клеить руками, но фокус группа из трех тимлидов не осилила. Это казалось не понятным, стремным, и не очевидным (вероятно нужно больше времени было потратить на UI/UX этой формы). Так же стал вопрос, как сохранить накликанные правила, ведь сервис не хранит данные и не отличает один проект от другого (хранить для всех пользователей?)

Все это время был механизм от git для этой цели. Файл .mailmap, который поддерживается много лет, имеет документацию и стандартизирован. Поэтому было ощущение, что решаем проблему НЕ нашей системы. Если пользователи будут заполнять этот .mailmap по правилам, то статистика будет корректно работать не только у нас, но и в любом другом из 20 аналогичных пакетов визуализации. Поэтому решили сослаться на него и удалить код.

**В данный момент** сделали два шага, чтобы помочь пользователю:
1. На странице настроек подготовили содержание .mailmap для копирования и редактирования. Т.к. скопировать и править готовый шаблон легче, чем писать с нуля.
2. Сегодня добавил склейку Имя+Email, которая будет править имя или почту, если до этого в логе уже была какая-то аналогичная пара. Пример:

![image](https://github.com/bakhirev/assayo/assets/2678592/82011b05-c871-4a97-a852-98c8001d7d43)


Не уверен что пункт 2 не создаст новых багов для людей, которые по какой-то причине сделали цепочку переименований Имя+Почта, Почта+Имя

**Comment by Igralino** [comment]
Спасибо за подробный ответ!
